File: src/lang/clj/clojure/set.clj
;; clojure.set — Phase 6.16.b-1 (.clj migration).
;;
;; Loaded by `src/lang/bootstrap.zig::loadCore` per ADR-0032 multi-file
;; FILES table. The Group A + B vars (`union` / `intersection` /
;; `difference` / `subset?` / `superset?` / `rename-keys` /
;; `map-invert`) are pure-Clojure Pattern A defns per ADR-0033 D3 + v5
;; §8.2. Each composes `reduce` / `conj` / `disj` / `contains?` /
;; `every?` / `assoc` / `dissoc` / `get` / `count` from rt/ — visible
;; unqualified here because `evalInNs` refers rt/ into the entered ns
;; (commit 6.16.b-1 + ADR-0035 in Phase 6.16.b-4 codifies this as a
;; proper `(ns ...)` macro).
;;
;; **Variadic via [& sets] + internal arity discrimination**: union /
;; intersection / difference accept 0/1/2/3+ args using a single
;; rest-arg `fn*` form (no multi-arity dispatch needed). This sidesteps
;; D-070 (multi-arity `fn*`) for these three vars — the survey's D-070
;; back-fill plan is therefore void for this group.
;;
;; Group C (`select` / `project` / `index` / `rename` / `join`) lands
;; at 6.16.b-3 after D-061 (`#{}` reader literal) + D-059 (map-literal
;; analyzer) infra ships in 6.16.b-2.

(in-ns 'clojure.set)

(def union
  (fn* [& sets]
    (if (= 0 (count sets))
      (hash-set)
      (reduce (fn* [acc s] (reduce conj acc s))
              (first sets)
              (rest sets)))))

(def intersection
  (fn* [& sets]
    (if (= 0 (count sets))
      nil
      (if (= 1 (count sets))
        (first sets)
        (reduce (fn* [s1 s2]
                  (reduce (fn* [acc x]
                            (if (contains? s2 x) acc (disj acc x)))
                          s1
                          s1))
                (first sets)
                (rest sets))))))

(def difference
  (fn* [& sets]
    (if (= 0 (count sets))
      nil
      (if (= 1 (count sets))
        (first sets)
        (reduce (fn* [s1 s2] (reduce disj s1 s2))
                (first sets)
                (rest sets))))))

(def subset?
  (fn* [s1 s2]
    (if (<= (count s1) (count s2))
      (every? (fn* [x] (contains? s2 x)) s1)
      false)))

(def superset?
  (fn* [s1 s2] (subset? s2 s1)))

;; `(rename-keys m kmap)` — rebuild m by replacing each old key in
;; kmap with its new-key partner. Skips entries whose old key is not
;; in m (matches JVM). The `(nth kv 0/1)` destructure substitutes
;; for vector binding inside `let*`.
(def rename-keys
  (fn* [m kmap]
    (reduce (fn* [acc kv]
              (let* [old (nth kv 0)
                     new-k (nth kv 1)]
                (if (contains? m old)
                  (assoc (dissoc acc old) new-k (get m old))
                  acc)))
            m
            kmap)))

;; `(map-invert m)` — swap keys and values. JVM uses a transient
;; reduce-kv for O(n) cost; cw v1 uses persistent reduce since
;; transients land at Phase 8 (DIVERGENCE D-α per per-task survey).
(def map-invert
  (fn* [m]
    (reduce (fn* [acc kv]
              (assoc acc (nth kv 1) (nth kv 0)))
            (hash-map)
            m)))

File: src/lang/primitive/set.zig
// SPDX-License-Identifier: EPL-2.0
//! `clojure.set/` namespace surface — Phase 6.16.b-1 (.clj migration).
//!
//! The Group A + B vars (`union` / `intersection` / `difference` /
//! `subset?` / `superset?` / `rename-keys` / `map-invert`) moved to
//! pure-Clojure Pattern A defns in `src/lang/clj/clojure/set.clj`
//! per ADR-0033 D3 + v5 §8.2. This file now ships only:
//!
//! - `hash-set` (variadic constructor; the `#{...}` reader literal
//!   lowers to a `hash-set` call until D-061 lands a `.set` Form
//!   variant + analyzer node at 6.16.b-2).
//! - `hash-map` (variadic constructor; same role for `{...}` until
//!   D-059 lands map-literal-as-Value at 6.16.b-2).
//!
//! Group C (`select` / `project` / `index` / `rename` / `join`) is
//! also a `.clj` defn — lands at 6.16.b-3 after the D-061 + D-059
//! infra ships.

const Value = @import("../../runtime/value/value.zig").Value;
const Runtime = @import("../../runtime/runtime.zig").Runtime;
const env_mod = @import("../../runtime/env.zig");
const Env = env_mod.Env;
const error_mod = @import("../../runtime/error/info.zig");
const SourceLocation = error_mod.SourceLocation;
const error_catalog = @import("../../runtime/error/catalog.zig");
const dispatch = @import("../../runtime/dispatch.zig");
const set_collection = @import("../../runtime/collection/set.zig");
const map_collection = @import("../../runtime/collection/map.zig");

/// `(hash-set & xs)` — construct a set from variadic args. Empty
/// arg list returns the empty-set singleton. Each arg is conj-ed
/// in order (idempotent — duplicates collapse).
pub fn hashSet(rt: *Runtime, env: *Env, args: []const Value, loc: SourceLocation) anyerror!Value {
    _ = env;
    _ = loc;
    var s = set_collection.empty();
    for (args) |a| s = try set_collection.conj(rt, s, a);
    return s;
}

/// `(hash-map & kvs)` — construct a map from variadic key/value pairs.
/// Odd argument count raises `map_literal_arity_odd` (matches the
/// JVM `IllegalArgumentException`). Empty arg list returns the
/// empty-map singleton.
pub fn hashMap(rt: *Runtime, env: *Env, args: []const Value, loc: SourceLocation) anyerror!Value {
    _ = env;
    if (args.len % 2 != 0)
        return error_catalog.raise(.map_literal_arity_odd, loc, .{});
    var m = map_collection.empty();
    var i: usize = 0;
    while (i < args.len) : (i += 2) {
        m = try map_collection.assoc(rt, m, args[i], args[i + 1]);
    }
    return m;
}

// --- registration ---

const Entry = struct {
    name: []const u8,
    f: dispatch.BuiltinFn,
};

const RT_ENTRIES = [_]Entry{
    .{ .name = "hash-set", .f = &hashSet },
    .{ .name = "hash-map", .f = &hashMap },
};

/// Register `hash-set` / `hash-map` into `rt/` (so they are user-
/// callable unqualified after `(refer 'rt)` into `user/`) and ensure
/// the `clojure.set` namespace exists for the .clj loader to enter.
/// The Group A + B vars themselves are registered by evaluating
/// `src/lang/clj/clojure/set.clj` at bootstrap.
pub fn register(env: *Env) !void {
    const rt_ns = env.findNs("rt") orelse return error.RtNamespaceMissing;
    for (RT_ENTRIES) |it| {
        _ = try env.intern(rt_ns, it.name, Value.initBuiltinFn(it.f), null);
    }
    _ = try env.findOrCreateNs("clojure.set");
}
